### ISSUE TITLE
[HOLD for payment 2024-12-25] [Hold for #53360] [$250] Company cards toggle is disabled after upgrading workspace on mobile

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#53505]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @joekaufmanexpensify 
**Slack conversation** (hyperlinked to channel name): [ts_external_expensify_expense](https://expensify.slack.com/archives/C06ML6X0W9L/p1733261806214079)

## Action Performed:
Prerequisite: Workspace created and on collect policy

1. Enable Company cards in more feature 
2. Click "upgrade" in the popup for Upgrade to control policy
3. Click Ok got it

## Expected Result:
Company cards toggle button in enabled condition as user upgraded to control policy
## Actual Result:
Company cards toggle button in disabled condition
## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [x] iOS: Standalone
- [x] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

https://github.com/user-attachments/assets/45ac1dcd-3bcf-4ca6-b126-f8be035eb1aa

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021864369617192885056</li>
        <li>Upwork Job ID: 1864369617192885056</li>
        <li>Last Price Increase: 2024-12-04</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sonialiap</details>

**Comment by ikevin127** [comment]
Not reproducible on **iOS: Native**, if this is reproducible only on **iOS: HybridApp** then it should be `Internal` because `External` contributors don't have access to the **HybridApp** repo.

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Company cards toggle is disabled after upgrading workspace on mobile

### What is the root cause of that problem?

This issue occurred in this pull request: https://github.com/Expensify/App/pull/49937.

After migrating E/App to `usePlatformStackNavigation`, the navigation blur event stopped working when navigating back to the Workspace More feature.

https://github.com/Expensify/App/blob/6a9854f089bd30c86d6b26d34055e77c1a7f95b0/src/pages/workspace/upgrade/WorkspaceUpgradePage.tsx#L159

For this reason, `confirmUpgrade` is not called, which has caused the company card toggle state to remain disabled.

https://github.com/Expensify/App/blob/6a9854f089bd30c86d6b26d34055e77c1a7f95b0/src/pages/workspace/upgrade/WorkspaceUpgradePage.tsx#L163

The screen is unmounted immediately when navigating away in case of native stack, so we won't get any events from https://github.com/react-navigation/react-navigation/issues/11788#issuecomment-1898186947

### What changes do you think we should make in order to solve the problem?

To resolve this issue, we should use `useFocusEffect` to handle the `confirmUpgrade` logic instead of listening for the blur event. Some thing like that:

1. Remove this block
https://github.com/Expensify/App/blob/6a9854f089bd30c86d6b26d34055e77c1a7f95b0/src/pages/workspace/upgrade/WorkspaceUpgradePage.tsx#L158-L167

2. Add `useFocusEffect` to handle the `confirmUpgrade` logic 
```diff
//src/pages/workspace/upgrade/WorkspaceUpgradePage.tsx#L158
    useFocusEffect(
        useCallback(() => {
            return () => {
                if (!isUpgraded || !canPerformUpgrade) {
                    return;
                }
                confirmUpgrade();
            };
        }, [isUpgraded, canPerformUpgrade, confirmUpgrade]),
    );
```

<details>
<summary>POC</summary>

https://github.com/user-attachments/assets/8c829909-1e1f-4cb1-a69e-e853062f6318


</details>
### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?


N/A

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by saifelance** [comment]
@dukenv0307 
You're absolutely right that if you're already using `navigation.addListener` in multiple places, switching to `useFocusEffect` might not be the best solution. Given that, it would be more appropriate to fix the root cause of why the navigation blur event doesn't work properly after migrating to `usePlatformStackNavigation`.

### Issue:
The `navigation.addListener('blur')` event isn't being triggered correctly after the migration to `usePlatformStackNavigation`. The root cause is likely due to screen unmounting immediately with native stack navigation, preventing the blur event from being captured.

### Solution:
Instead of replacing `navigation.addListener` with `useFocusEffect`, we can ensure that the navigation events are properly captured by correctly managing the screen lifecycle with the native stack. Specifically, we can use the `navigation.addListener('beforeRemove')` event which gets triggered before the screen is removed from the navigation stack.

### Key to Fix:

1. Keep `navigation.addListener('blur')`: We’ll fix the issue by keeping navigation.addListener but ensuring that the event is triggered at the correct time.
2. Switch to `beforeRemove` event: Use `beforeRemove` to capture navigation changes before the screen is unmounted.

Here's an updated version of the `useEffect` with `beforeRemove`:


```
useEffect(() => {
    const unsubscribeListener = navigation.addListener('beforeRemove', () => {
        if (!isUpgraded || !canPerformUpgrade) {
            return;
        }
        confirmUpgrade();
    });

    return unsubscribeListener;
}, [isUpgraded, canPerformUpgrade, confirmUpgrade, navigation]);

```

**Comment by huult** [comment]
@dukenv0307 Firstly, I would like to say thank you for reviewing my proposal

> Thank you @huult, but I don't think it's the best idea since we're using navigation.addListener in many places.

Since this issue only happens on the Native Stack, it means it will occur on both native Android and iOS. I believe the issue might also happen elsewhere. Therefore, I suggest creating a custom hook,ex: useOnBlur, with two files: index.ts and index.native.ts, to handle native and web platforms. This approach will make it easier to apply the solution elsewhere.

> It would be great if we can fix this issue

As mentioned in this comment https://github.com/react-navigation/react-navigation/issues/11788#issuecomment-1898186947, they confirmed that addListener with blur does not work on the Native Stack, so we can't fix it. They suggested using useFocusEffect to handle this issue. Therefore, I think we should use useFocusEffect for native platforms and blur for the web.

**Comment by CyberAndrii** [comment]
Dupe of https://github.com/Expensify/App/issues/53360...

**Comment by sonialiap** [comment]
Putting this on hold for https://github.com/Expensify/App/issues/53360

**Comment by sonialiap** [comment]
https://github.com/Expensify/App/pull/53735 resolved the issue, payment is in https://github.com/Expensify/App/issues/53360. Closing this out
